cub/cub/device/device_transform.cuh

```diff
@@ -0,0 +1,230 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#include <cub/config.cuh>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cub/detail/nvtx.cuh>
+#include <cub/device/dispatch/dispatch_transform.cuh>
+#include <cub/util_namespace.cuh>
+
+#include <cuda/std/tuple>
+
+CUB_NAMESPACE_BEGIN
+
+//! DeviceTransform provides device-wide, parallel operations for transforming elements tuple-wise from multiple input
+//! streams into an output stream.
+struct DeviceTransform
+{
+  // Many input streams, one output stream
+
+  /// @rst
+  /// Overview
+  /// +++++++++++++++++++++++++++++++++++++++++++++
+  /// Transforms many input streams into one output stream, by applying a transformation operation on corresponding
+  /// input elements and writing the result to the corresponding output element. No guarantee is given on the identity
+  /// (i.e. address) of the objects passed to the call operator of the transformation operation.
+  ///
+  /// A Simple Example
+  /// +++++++++++++++++++++++++++++++++++++++++++++
+  ///
+  /// .. literalinclude:: ../../../cub/test/catch2_test_device_transform_api.cu
+  ///     :language: c++
+  ///     :dedent:
+  ///     :start-after: example-begin transform-many
+  ///     :end-before: example-end transform-many
+  ///
+  /// @endrst
+  ///
+  /// @param count The number of elements in each input stream.
+  /// @param inputs A tuple of iterators to the input streams where count elements are read from each. The iterators'
+  /// value types must be trivially relocatable.
+  /// @param output An iterator to the output stream where count results are written to.
+  /// @param transform_op An n-ary function object, where n is the number of input streams. The input iterators' value
+  /// types must be convertible to the parameters of the function object's call operator. The return type of the call
+  /// operator must be assignable to the dereferenced output iterator.
+  template <typename... RandomAccessIteratorsIn, typename RandomAccessIteratorOut, typename TransformOp>
+  CUB_RUNTIME_FUNCTION static cudaError_t Transform(
+    int count,
```

Comment:
important: let's use `num_items` consistently

Reply:
question: can we accept templated `num_items` and rely on `choose_offset_t` here?

Reply:
You told me in the past to take counts as `int` until we know how to deal with large offsets in the CUB API. I added it as task to #1947 and would like to address it after this PR is merged.